src/main/java/io/github/thebusybiscuit/slimefun4/storage/backend/legacy/LegacyStorage.java

```diff
@@ -0,0 +1,116 @@
+
+import com.google.common.annotations.Beta;
+
+import javax.annotation.Nonnull;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.UUID;
+import java.util.logging.Level;
+
+@Beta
+public class LegacyStorage implements Storage {
+
+    @Override
+    public PlayerData loadPlayerData(@Nonnull UUID uuid) {
+        Config playerFile = new Config("data-storage/Slimefun/Players/" + uuid + ".yml");
+        // Not too sure why this is it's own file
+        Config waypointsFile = new Config("data-storage/Slimefun/waypoints/" + uuid + ".yml");
+
+        // Load research
+        Set<Research> researches = new HashSet<>();
+        for (Research research : Slimefun.getRegistry().getResearches()) {
+            if (playerFile.contains("researches." + research.getID())) {
+                researches.add(research);
+            }
+        }
+
+        // Load backpacks
+        HashMap<Integer, PlayerBackpack> backpacks = new HashMap<>();
+        for (String key : playerFile.getKeys("backpacks")) {
+            try {
+                int id = Integer.parseInt(key);
+                int size = playerFile.getInt("backpacks." + key + ".size");
+
+                HashMap<Integer, ItemStack> items = new HashMap<>();
+                for (int i = 0; i < size; i++) {
+                    items.put(i, playerFile.getItem("backpacks." + key + ".contents." + i));
+                }
+
+                PlayerBackpack backpack = PlayerBackpack.load(uuid, id, size, items);
+
+                backpacks.put(id, backpack);
+            } catch (Exception x) {
+                Slimefun.logger().log(Level.WARNING, x, () -> "Could not load Backpack \"" + key + "\" for Player \"" + uuid + '"');
+            }
+        }
+
+        // Load waypoints
+        Set<Waypoint> waypoints = new HashSet<>();
+        for (String key : waypointsFile.getKeys()) {
+            try {
+                if (waypointsFile.contains(key + ".world") && Bukkit.getWorld(waypointsFile.getString(key + ".world")) != null) {
+                    String waypointName = waypointsFile.getString(key + ".name");
+                    Location loc = waypointsFile.getLocation(key);
+                    waypoints.add(new Waypoint(uuid, key, loc, waypointName));
+                }
+            } catch (Exception x) {
+                Slimefun.logger().log(Level.WARNING, x, () -> "Could not load Waypoint \"" + key + "\" for Player \"" + uuid + '"');
+            }
+        }
+
+        return new PlayerData(researches, backpacks, waypoints);
+    }
+
+    @Override
+    public void savePlayerData(@Nonnull UUID uuid, @Nonnull PlayerData data) {
+        Config playerFile = new Config("data-storage/Slimefun/Players/" + uuid + ".yml");
+        // Not too sure why this is it's own file
+        Config waypointsFile = new Config("data-storage/Slimefun/waypoints/" + uuid + ".yml");
```

Comment:
Should we extract these and make them class variables? They are used in both load and save, and the first one is still referenced in PlayerProfile.java just for a getter that I see, and who knows if they are elsewhere that I have yet to see or can't remember. Sounds more robust to just have them be in one place, or is there a reason to keep them elsewhere too?

Reply:
Bit awkward since they rely on a UUID. I could put the string as a global and format but it's only used twice in here so wasn't the biggest fan of that.
This API isn't great and not really expandable so I plan to work on figuring out how that should look post this PR. Once that is done, we may be able to globalise in some way. Not sure how yet.